packages/repository/src/relations/has-many/has-many-through.repository.ts

```diff
@@ -0,0 +1,151 @@
+import {Entity} from '../../model';
+import {Filter, Where} from '../../query';
+import {
+  constrainDataObject,
+  constrainFilter,
+  constrainWhere,
+} from '../../repositories/constraint-utils';
+import {EntityCrudRepository} from '../../repositories/repository';
+import {AdvancedConstraint} from './has-many-through-repository.factory';
+
+/**
+ * CRUD operations for a target repository of a HasMany relation
+ */
+export interface HasManyThroughRepository<
+  Target extends Entity,
+  Through extends Entity
+> {
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param throughModelData The through model data
+   * @param options Options for the operation
+   * @param throughOptions Options passed to create through
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(
+    targetModelData: DataObject<Target>,
+    throughModelData?: DataObject<Through>,
+    options?: Options,
+    throughOptions?: Options,
+  ): Promise<Target>;
+  /**
+   * Find target model instance(s)
+   * @param filter A filter object for where, order, limit, etc.
+   * @param options Options for the operation
+   * @returns A promise which resolves with the found target instance(s)
+   */
+  find(filter?: Filter<Target>, options?: Options): Promise<Target[]>;
+  /**
+   * Delete multiple target model instances
+   * @param where Instances within the where scope are deleted
+   * @param options
+   * @returns A promise which resolves the deleted target model instances
+   */
+  delete(where?: Where<Target>, options?: Options): Promise<Count>;
+  /**
+   * Patch multiple target model instances
+   * @param dataObject The fields and their new values to patch
+   * @param where Instances within the where scope are patched
+   * @param options
+   * @returns A promise which resolves the patched target model instances
+   */
+  patch(
+    dataObject: DataObject<Target>,
+    where?: Where<Target>,
+    options?: Options,
+  ): Promise<Count>;
+}
+
+export class DefaultHasManyThroughRepository<
+  TargetEntity extends Entity,
+  TargetID,
+  TargetRepository extends EntityCrudRepository<TargetEntity, TargetID>,
+  ThroughEntity extends Entity,
+  ThroughID,
+  ThroughRepository extends EntityCrudRepository<ThroughEntity, ThroughID>
+> implements HasManyThroughRepository<TargetEntity, ThroughEntity> {
+  /**
+   * Constructor of DefaultHasManyEntityCrudRepository
+   * @param getTargetRepository the getter of the related target model repository instance
+   * @param constraint the key value pair representing foreign key name to constrain
+   * the target repository instance
+   */
+  constructor(
+    public getTargetRepository: Getter<TargetRepository>,
+    public getThroughRepository: Getter<ThroughRepository>,
+    public getAdvancedConstraint: (
+      targetInstance?: TargetEntity,
+    ) => Promise<AdvancedConstraint<TargetEntity | ThroughEntity>>,
```

Comment:
As commented above, I think the constraint type should be parameterized by both "target" and "through" classes, and use a more descriptive name.

```suggestion
    ) => Promise<HasManyThroughConstraint<TargetEntity, ThroughEntity>>,
```

Reply:
👍 will do

Reply:
Done :)

Take a look at the following comment.

https://github.com/strongloop/loopback-next/pull/2359#issuecomment-475979639